designs/carbon-aware.md

```diff
@@ -0,0 +1,197 @@
+-	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster) *Controller {
++func NewController(ctx context.Context, clk clock.Clock, kubeClient client.Client, provisioner *provisioning.Provisioner,
++	cp cloudprovider.CloudProvider, recorder events.Recorder, cluster *state.Cluster) *Controller {
+
++	if settings.FromContext(ctx).CarbonEfficientEnabled {
++		return &Controller{
++			clock:         clk,
++			kubeClient:    kubeClient,
++			cluster:       cluster,
++			provisioner:   provisioner,
++			recorder:      recorder,
++			cloudProvider: cp,
++			lastRun:       map[string]time.Time{},
++			deprovisioners: []Deprovisioner{
++				NewExpiration(clk, kubeClient, cluster, provisioner, recorder),
++				NewDrift(kubeClient, cluster, provisioner, recorder),
++				NewEmptiness(clk),
++				NewEmptyMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
++				NewCarbonEfficientMultiMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
++				NewCarbonEfficientSingleMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
++			},
++		}
++	}
+
+	return &Controller{
+		clock:         clk,
+		kubeClient:    kubeClient,
+		cluster:       cluster,
+		provisioner:   provisioner,
+		recorder:      recorder,
+		cloudProvider: cp,
+		lastRun:       map[string]time.Time{},
+		deprovisioners: []Deprovisioner{
+			NewExpiration(clk, kubeClient, cluster, provisioner, recorder),
+			NewDrift(kubeClient, cluster, provisioner, recorder),
+			NewEmptiness(clk),
+			NewEmptyMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+			NewMultiMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+			NewSingleMachineConsolidation(clk, cluster, kubeClient, provisioner, cp, recorder),
+		},
+	}
+}
+```
+</details>
+
+#### Provisioning
+In `karpenter-core`, create a new method `types.go/OrderByCarbonEmissions` and use that in `nodeclaimtemplate.go/ToMachine` and `nodeclaimtemplate.go/ToNodeClaim` instead of `types.go/OrderByPrice` when Carbon Efficient is enabled.
+
+In `karpenter`, create a new method `CarbonEfficientCreate` in `pkg/providers/instance/instance.go` that is used in `pkg/cloudprovider/cloudprovider.go/Create` instead of `pkg/providers/instance/instance.go/Create` when Carbon Efficient is enabled.
+
+#### Considerations
+1. 👍 Current consolidation methods are unaffected.
+1. 👎 There might be copy-paste of code from the original consolidation methods to the carbon efficient consolidators.
+
+### Option 2: Use Carbon Efficient filtering/sorting methods
+
+#### Consolidation
+Create carbon efficient implementations of low-level functions like `filterByPrice`, `filterOutSameType`, `getCandidatePrices`, etc. that are used when Carbon Efficient is enabled. Usage of aforementioned functions might assume that it is price that they are getting, but in reality it is data about carbon emissions.
```

Comment:
Can you go into further depth where the data consolidation will use or consider? How do cloud providers provide this carbon aware data to karpenter core?